### DISCUSSION TITLE
Export from UML2, possibility to use xtend?

### DISCUSSION THREAD

**Discussion by ansgarradermacher** [hallvard/plantuml#178] [General]
Dear all,
I'm currently prototyping a PlantUML export from UML2 (Papyrus) models. There is already support for sequence, class, and state-machine diagrams. The code would be smaller and easier to maintain if it could profit from text templates available in xtend (e.g. avoid String concatenation and have good indentation). But this would add a dependency for building and execution. Would that be fine? (we haven't decided yet, whether to provide the export within Papyrus or the PlantUML Eclipse plugin).

**Comment by travkin79** [comment]
Hello @ansgarradermacher,
Thank you for working on that export feature.

Concerning the use of xtend: I prefer not to add an xtend dependency to PlantUML Eclipse plug-ins. I think, in general, there are benefits in limiting the number of dependencies to other projects to a minimum has maintenance benefits. In case of xtend, I'd like to avoid having a dependency to a project that is not being maintained any longer, [the repo has been archived](https://github.com/eclipse/xtext-xtend) and it seems, its maintenance has stopped (see [related discussion](https://github.com/eclipse/xtext/issues/1721)).

In case you would like to use xtext (maybe with xpand), not xtend, I'd suggest to add a PlantUML feature with one or multiple plug-ins that use xtext and avoid adding an xtext dependency to the PlantUML core plug-ins. This way, a user could decide to only use a core version of PlantUML tooling in Eclipse, without having to install xtext.

If you plan to implement the export in Papyrus, feel free to do whatever you think makes sense. :-)

**Reply by OepenerPlatt** [reply]
Hello @travkin79 and @ansgarradermacher,
correct me if I'm wrong, but I understood that the [xtend code was moved into the xtext codebase](https://github.com/eclipse/xtext-xtend?tab=readme-ov-file#notice) for easier maintenance. 
The [eclipse marketplace is still maintained too](https://marketplace.eclipse.org/content/eclipse-xtend) (Date Updated: Tuesday, February 27, 2024 - 05:31)
I use both xcore and xtend a lot since many years. 
It would be a pity if these superb tools would not be available anymore.

**Reply by travkin79** [reply]
Hi @OepenerPlatt,
You're right. The xtend source code was moved to the xtext mono-repo, but that was done already [8 years ago](https://github.com/eclipse/xtext/commit/b67f8cb766c9ffbae390b0d979c7bd5a21675188). The maintenance problems came later.

It seems, there are still recent xtext and xtend releases, but there are severe xtend maintenance problems due to missing resources (see [release notes](https://eclipse.dev/Xtext/xtend/releasenotes.html) and the linked discussion on [xtext's and xtend's future](https://github.com/eclipse/xtext/issues/1721)). There is a discussion where the people tend to stop maintaining xtend in favor of xtext.
![image](https://github.com/hallvard/plantuml/assets/10887297/3eb26a71-1698-4e41-8619-7283295989c0)

I know of some teams that are already removing xtend from their projects, but still relying on xtext.

**Reply by OepenerPlatt** [reply]
Hi @travkin79,
this should be an incentive for me to try to contribute to the xtend project in order to keep the language alive.
Thanks for your input.

**Comment by ansgarradermacher** [comment]
Hello @travkin79,
thanks for the reminder that the future of xtend is indeed not sure. We will do without xtend for the PlantUML export - we will have some effort in Papyrus to migrate, once xtend is no longer in the Eclipse release train. The use of xtext with direct Java generation is an option for xtext-based editors, but does not replace the xtend code in the code generators (as far as I know).
